core/services/vrf/listener_v2.go

```diff
@@ -306,11 +348,234 @@ func (a fulfilledReqV2) Compare(b heaps.Item) int {
+
+			// This check to see if the log was consumed needs to be in the same
+			// goroutine as the mark consumed to avoid processing duplicates.
+			consumed, errConsumed := lsn.logBroadcaster.WasAlreadyConsumed(req.lb)
+			if errConsumed != nil {
+				// Do not process for now, retry on next iteration.
+				rlog.Errorw("Could not determine if log was already consumed", "err", errConsumed)
+				continue
+			} else if consumed {
+				processed[vrfRequest.RequestId.String()] = struct{}{}
+				continue
+			}
+
+			// Check if we can ignore the request due to it's age.
+			if time.Now().UTC().Sub(req.utcTimestamp) >= lsn.job.VRFSpec.RequestTimeout {
+				rlog.Infow("Request too old, dropping it")
+				lsn.markLogAsConsumed(req.lb)
+				processed[vrfRequest.RequestId.String()] = struct{}{}
+				incDroppedReqs(lsn.job.Name.ValueOrZero(), lsn.job.ExternalJobID, v2, reasonAge)
+				continue
+			}
+
+			// Check if the vrf req has already been fulfilled.
+			// Pass max(request.BlockNumber, currentBlockNumber) to the call to ensure we get
+			// the latest state or an error.
+			// This is to avoid a race condition where a request commitment has been written on-chain
+			// but has not been synced fully on the entire network. GetCommitment will return an empty
+			// byte array in that case, and we would think that the request is fulfilled when it actually
+			// isn't.
+			reqBlockNumber := new(big.Int).SetUint64(vrfRequest.Raw.BlockNumber)
+			currBlock := new(big.Int).SetUint64(lsn.getLatestHead())
+			m := bigmath.Max(reqBlockNumber, currBlock)
+			callback, errCb := lsn.coordinator.GetCommitment(&bind.CallOpts{
+				BlockNumber: m,
+			}, vrfRequest.RequestId)
+			if errCb != nil {
+				rlog.Errorw("Unable to check if already fulfilled, processing anyways", "err", errCb)
+			} else if utils.IsEmpty(callback[:]) {
+				// If seedAndBlockNumber is zero then the response has been fulfilled
+				// and we should skip it
+				rlog.Infow("Request already fulfilled", "callback", callback)
+				lsn.markLogAsConsumed(req.lb)
+				processed[vrfRequest.RequestId.String()] = struct{}{}
+				continue
+			}
+			// Run the pipeline to determine the max link that could be billed at maxGasPrice.
+			// The ethcall will error if there is currently insufficient balance onchain.
+			maxLink, run, _, gaslimit, pf, rc, errG := lsn.getMaxLinkForFulfillment(maxGasPriceWei, req, rlog)
+			if errG != nil {
+				rlog.Warnw("Unable to get max link for fulfillment, skipping request", "err", errG)
+				continue
+			}
+			if startBalanceNoReserveLink.Cmp(maxLink) < 0 {
+				// Insufficient funds, have to wait for a user top up
+				// leave it unprocessed for now
+				rlog.Infow("Insufficient link balance to fulfill a request, breaking", "maxLink", maxLink)
+				break
+			}
+
+			// check if we've exceeded the block gas limit.
+			// if we have, we should break here and re-process this request in the next
+			// batch.
+			if (totalGasLimit + gaslimit) > blockGasLimit {
```

Comment:
Yeah I think you're right, this is a bit tricky.

Reply:
We can pull the max callback gas limit from the contract but one concern I have is for getting these batch fulfillments included - the higher the gas limit on the fulfillment tx the less likely miners/validators are keen to include it, since the predominant strategy is to include as many low-gas but high-fee tx's as possible within a single block.

Reply:
But anyway, I think the strategy of using the max callback limit multiplied by some multiplier or + some buffer is better than these arbitrary numbers I have here.